### PR TITLE
DOC: Added missing C-API functions

### PR DESCRIPTION
**Pull request by Mukulikaa**
Added the following functions to NumPy C-API:
1. PyArray_CopyObject
2. PyArray_MapIterArray
3. PyArray_MapIterSwapAxes
4. PyArray_MapIterNext
5. PyArray_MapIterArrayCopyIfOverlap

### CODE REVIEW

**Comment by melissawm** [comment]
[Rendered artifact here](https://21998-908607-gh.circle-artifacts.com/0/doc/build/html/reference/c-api/array.html)

File: doc/source/reference/c-api/array.rst [hunk 1]
```diff
@@ -723,6 +724,13 @@ From other objects
     broadcastable to the shape of ``dest``. The data areas of dest
     and src must not overlap.
 
+.. c:function:: int PyArray_CopyObject(PyArrayObject* dest, PyObject* src)
+
+    Assign an object ``src`` to a NumPy array ``dest`` according to
+    array-coercion rules. This is largely basically identical to
```
Comment (melissawm) [inline doc/source/reference/c-api/array.rst:730]:
```suggestion
    array-coercion rules. This is basically identical to
```

**Review by melissawm — COMMENTED** [review]
I can't review the technical details, so I'll leave that for others to look at. The only question I have is if the `core/include/numpy/multiarray_api.txt` file also needs to be updated to include the "docstring" for `PyArray_CopyObject`- I'm not sure what is the function of that file.

File: doc/source/reference/c-api/array.rst [hunk 2]
```diff
@@ -584,25 +584,26 @@ From other objects
     did not have the _ARRAY_ macro namespace in them. That form
     of the constant names is deprecated in 1.7.
 
-.. c:macro:: NPY_ARRAY_NOTSWAPPED
+    .. c:macro:: NPY_ARRAY_NOTSWAPPED
```
Comment (mattip) [inline doc/source/reference/c-api/array.rst:587]: These are un-indented on purpose so that the macros can be linked to from elsewhere

**Comment by mattip** [comment]
This seems like a good opportunity to use the new [doxygen-breathe infrastructure](https://numpy.org/devdocs/dev/howto-docs.html?highlight=doxygen#documenting-c-c-code) to pull the comments out of the source files and render them in the rst files

**Comment by Mukulikaa** [comment]
I'd be happy to try it out then!

**Comment by rgommers** [comment]
> This seems like a good opportunity to use the new doxygen-breathe infrastructure to pull the comments out of the source files and render them in the rst files

I'd suggest merging this once content seems ready, and using Doxygen-Breathe in a follow-up PR to not mix content and tooling changes.

**Comment by rossbar** [comment]
> using Doxygen-Breathe in a follow-up PR to not mix content and tooling changes.

FWIW the tooling changes have already been merged, so the proposal is to use this as a test-case for the C autodocumentation workflow. @Mukulikaa has expressed interest (or at least willingness) to give this a try and provide feedback on how the process goes. @rgommers brings up a good point though - if you get stuck don't feel the need to keep pushing on the autodocumenting "experiment", that shouldn't be a blocker for the content itself.

**Comment by charris** [comment]
Let's get this in. Thanks @Mukulikaa .

**Pull request merged**